iati/validator.py

```diff
@@ -282,43 +343,37 @@ def _check_codes(dataset, codelist):
     Returns:
         iati.validator.ValidationErrorLog: A log of the errors that occurred.
 
+    Raises:
+        ValueError: When a path in a mapping is not looking for an attribute value or element text.
+
     """
     error_log = ValidationErrorLog()
     mappings = iati.default.codelist_mapping()
 
     for mapping in mappings[codelist.name]:
-        base_xpath = mapping['xpath']
-        condition = mapping['condition']
-        split_xpath = base_xpath.split('/')
-        parent_el_xpath = '/'.join(split_xpath[:-1])
-        attr_name = split_xpath[-1:][0][1:]
-
-        if condition is None:
-            parent_el_xpath = parent_el_xpath + '[@' + attr_name + ']'
-        else:
-            parent_el_xpath = parent_el_xpath + '[' + condition + ' and @' + attr_name + ']'
-
-        # some nasty string manipulation to make the `//@xml:lang` mapping work
-        while not parent_el_xpath.startswith('//'):
-            parent_el_xpath = '/' + parent_el_xpath
-        if parent_el_xpath.startswith('//['):
-            parent_el_xpath = '//*[' + parent_el_xpath[3:]
-        # provide a secondary cludge to deal with the 'xml' namespace
-        if attr_name == 'xml:lang':
-            attr_name = '{http://www.w3.org/XML/1998/namespace}lang'
+        parent_el_xpath, last_xpath_section = mapping['xpath'].rsplit('/', 1)
 
-        parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
-
-        for parent in parents_to_check:
-            code = parent.attrib[attr_name]
+        if last_xpath_section.startswith('@'):
+            attr_name = last_xpath_section[1:]  # also used via `locals()` - DO NOT REMOVE THIS VAR
+            located_codes = _extract_codes_from_attrib(dataset, parent_el_xpath, attr_name, mapping['condition'])
+        elif last_xpath_section == 'text()':
+            located_codes = _extract_codes_from_element_text(dataset, parent_el_xpath, mapping['condition'])
+        else:
+            raise ValueError('mapping path does not locate attribute value or element text')
 
+        for (code, line_number) in located_codes:  # `line_number` used via `locals()` # pylint: disable=unused-variable
             if code not in codelist.codes:
-                line_number = parent.sourceline  # used via `locals()` # pylint: disable=unused-variable
-
-                if codelist.complete:
-                    error = ValidationError('err-code-not-on-codelist', locals())
+                if last_xpath_section.startswith('@'):
+                    if codelist.complete:
```

Comment:
extract setting of error name to a variable

Reply:
c2df432c4a955e99eeb8c83465bd6c47fa041f8e